File: lib/web_ui/lib/src/engine/skwasm/skwasm_impl/src/paint.dart
import 'dart:wasm';

import 'package:ui/ui.dart' as ui;
import 'raw/raw_paint.dart';

class SkwasmPaint implements ui.Paint {
  factory SkwasmPaint() {
    return SkwasmPaint._fromHandle(paint_create());
  }

  SkwasmPaint._fromHandle(this._handle);
  PaintHandle _handle;

  PaintHandle get handle => _handle;

  ui.BlendMode _cachedBlendMode = ui.BlendMode.srcOver;

  @override
  ui.BlendMode get blendMode {
    return _cachedBlendMode;
  }

  @override
  set blendMode(ui.BlendMode blendMode) {
    if (_cachedBlendMode != blendMode) {
      _cachedBlendMode = blendMode;
      paint_setBlendMode(_handle, blendMode.index.toWasmI32());
    }
  }

  @override
  ui.PaintingStyle get style {
    return ui.PaintingStyle.values[paint_getPaintStyle(_handle).toIntSigned()];
  }

  @override
  set style(ui.PaintingStyle style) {
    paint_setPaintStyle(_handle, style.index.toWasmI32());
  }

  @override
  double get strokeWidth {
    return paint_getStrokeWidth(_handle).toDouble();
  }

  @override
  set strokeWidth(double width) {
    paint_setStrokeWidth(_handle, width.toWasmF32());
  }

  @override
  ui.StrokeCap get strokeCap {
    return ui.StrokeCap.values[paint_getStrokeCap(_handle).toIntSigned()];
  }

  @override
  set strokeCap(ui.StrokeCap cap) {
    paint_setStrokeCap(_handle, cap.index.toWasmI32());
  }

  @override
  ui.StrokeJoin get strokeJoin {
    return ui.StrokeJoin.values[paint_getStrokeJoin(_handle).toIntSigned()];
  }

  @override
  set strokeJoin(ui.StrokeJoin join) {
    paint_setStrokeJoin(_handle, join.index.toWasmI32());
  }

  @override
  bool get isAntiAlias {
    return paint_getAntiAlias(_handle).toIntSigned() != 0;
  }
  @override
  set isAntiAlias(bool value) {
    paint_setAntiAlias(_handle, value ? 1.toWasmI32() : 0.toWasmI32());
  }

  @override
  ui.Color get color {
    return ui.Color(paint_getColorInt(_handle).toIntSigned());
  }

  @override
  set color(ui.Color color) {
    paint_setColorInt(_handle, color.value.toWasmI32());
  }
  
  @override
  ui.ColorFilter? colorFilter;
  
  @override
  ui.FilterQuality filterQuality = ui.FilterQuality.none;
  
  @override
  ui.ImageFilter? imageFilter;
  
  @override
  bool invertColors = false;
  
  @override
  ui.MaskFilter? maskFilter;
  
  @override
  ui.Shader? shader;
  
  @override
  double strokeMiterLimit = 0.0;  
}

File: lib/web_ui/skwasm/paint.cpp
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "export.h"
#include "helpers.h"
#include "render_strategy.h"

using namespace Skwasm;

SKWASM_EXPORT Paint* paint_create() {
  auto paint = new Paint();

  // Antialias defaults to true in flutter.
  paint->setAntiAlias(true);
  return paint;
}

SKWASM_EXPORT void paint_dispose(Paint* paint) {
  delete paint;
}

SKWASM_EXPORT void paint_setBlendMode(Paint* paint, BlendMode mode) {
  paint->setBlendMode(mode);
}

// No getter for blend mode, as it's non trivial. Cache on the dart side.

SKWASM_EXPORT void paint_setStyle(Paint* paint, Paint::Style style) {
  paint->setStyle(style);
}

SKWASM_EXPORT Paint::Style paint_getStyle(Paint* paint) {
  return paint->getStyle();
}

SKWASM_EXPORT void paint_setStrokeWidth(Paint* paint, Scalar width) {
  paint->setStrokeWidth(width);
}

SKWASM_EXPORT Scalar paint_getStrokeWidth(Paint* paint) {
  return paint->getStrokeWidth();
}

SKWASM_EXPORT void paint_setStrokeCap(Paint* paint, Paint::Cap cap) {
  paint->setStrokeCap(cap);
}

SKWASM_EXPORT Paint::Cap paint_getStrokeCap(Paint* paint) {
  return paint->getStrokeCap();
}

SKWASM_EXPORT void paint_setStrokeJoin(Paint* paint, Paint::Join join) {
  paint->setStrokeJoin(join);
}

SKWASM_EXPORT Paint::Join paint_getStrokeJoin(Paint* paint) {
  return paint->getStrokeJoin();
}

SKWASM_EXPORT void paint_setAntiAlias(Paint* paint, bool antiAlias) {
  paint->setAntiAlias(antiAlias);
}

SKWASM_EXPORT bool paint_getAntiAlias(Paint* paint) {
  return paint->isAntiAlias();
}

SKWASM_EXPORT void paint_setColorInt(Paint* paint, Color colorInt) {
  paint->setColor(colorInt);
}

SKWASM_EXPORT Color paint_getColorInt(Paint* paint) {
  return paint->getColor();
}

SKWASM_EXPORT void paint_setMiterLimit(Paint* paint, Scalar miterLimit) {
  paint->setStrokeMiter(miterLimit);
}

SKWASM_EXPORT Scalar paint_getMiterLimit(Paint* paint) {
  return paint->getStrokeMiter();
}

SKWASM_EXPORT void paint_setShader(Paint* paint, SkShader* shader) {
  paint->setShader(sk_ref_sp<SkShader>(shader));
}

SKWASM_EXPORT void paint_setImageFilter(Paint* paint, ImageFilter* filter) {
  //paint->setImageFilter(sk_ref_sp<ImageFilter>(filter));
}

SKWASM_EXPORT void paint_setColorFilter(Paint* paint, ColorFilter* filter) {
  //paint->setColorFilter(sk_ref_sp<ColorFilter>(filter));
}

SKWASM_EXPORT void paint_setMaskFilter(Paint* paint, MaskFilter* filter) {
  //paint->setMaskFilter(sk_ref_sp<MaskFilter>(filter));
}
